File: packages/renderer/src/critic.js
/**
 * Inline diff via CriticMarkup (http://criticmarkup.com/):
 *   {++insert++}  {--delete--}  {~~old~>new~~}  {==highlight==}  {>>comment<<}
 * Content is treated as plain text and HTML-escaped on output.
 */

const OPEN = 0x7b; // {

const KINDS = {
  '++': { kind: 'add', close: '++}' },
  '--': { kind: 'del', close: '--}' },
  '~~': { kind: 'sub', close: '~~}' },
  '==': { kind: 'mark', close: '==}' },
  '>>': { kind: 'comment', close: '<<}' },
};

function criticRule(state, silent) {
  const start = state.pos;
  const src = state.src;
  if (src.charCodeAt(start) !== OPEN) return false;

  const spec = KINDS[src.slice(start + 1, start + 3)];
  if (!spec) return false;

  const contentStart = start + 3;
  const closeIdx = src.indexOf(spec.close, contentStart);
  if (closeIdx === -1 || closeIdx + spec.close.length > state.posMax) return false;

  const raw = src.slice(contentStart, closeIdx);

  if (!silent) {
    const token = state.push('cm_critic', '', 0);
    if (spec.kind === 'sub') {
      const cut = raw.indexOf('~>');
      token.meta = {
        kind: 'sub',
        old: cut === -1 ? raw : raw.slice(0, cut),
        neu: cut === -1 ? '' : raw.slice(cut + 2),
      };
    } else {
      token.meta = { kind: spec.kind, content: raw };
    }
  }

  state.pos = closeIdx + spec.close.length;
  return true;
}

export default function criticPlugin(md) {
  md.inline.ruler.before('emphasis', 'cm_critic', criticRule);
  const esc = md.utils.escapeHtml;

  md.renderer.rules.cm_critic = (tokens, idx) => {
    const meta = tokens[idx].meta;
    switch (meta.kind) {
      case 'add':
        return `<ins class="crit-add">${esc(meta.content)}</ins>`;
      case 'del':
        return `<del class="crit-del">${esc(meta.content)}</del>`;
      case 'sub':
        return `<del class="crit-del">${esc(meta.old)}</del><ins class="crit-add">${esc(meta.neu)}</ins>`;
      case 'mark':
        return `<mark class="crit-mark">${esc(meta.content)}</mark>`;
      case 'comment':
        return `<span class="crit-comment">${esc(meta.content)}</span>`;
      default:
        return '';
    }
  };
}

File: packages/python/src/chromamark/critic.py
"""Inline diff via CriticMarkup: {++add++} {--del--} {~~old~>new~~} {==mark==} {>>comment<<}."""

from markdown_it.common.utils import escapeHtml

KINDS = {
    "++": ("add", "++}"),
    "--": ("del", "--}"),
    "~~": ("sub", "~~}"),
    "==": ("mark", "==}"),
    ">>": ("comment", "<<}"),
}


def _rule(state, silent):
    start = state.pos
    src = state.src
    if start >= state.posMax or src[start] != "{":
        return False
    spec = KINDS.get(src[start + 1:start + 3])
    if not spec:
        return False
    kind, close = spec

    content_start = start + 3
    close_idx = src.find(close, content_start)
    if close_idx == -1 or close_idx + len(close) > state.posMax:
        return False

    raw = src[content_start:close_idx]
    if not silent:
        tok = state.push("cm_critic", "", 0)
        if kind == "sub":
            cut = raw.find("~>")
            tok.meta = {
                "kind": "sub",
                "old": raw if cut == -1 else raw[:cut],
                "neu": "" if cut == -1 else raw[cut + 2:],
            }
        else:
            tok.meta = {"kind": kind, "content": raw}

    state.pos = close_idx + len(close)
    return True


def critic_plugin(md):
    md.inline.ruler.before("emphasis", "cm_critic", _rule)

    def render_critic(self, tokens, idx, options, env):
        meta = tokens[idx].meta
        kind = meta["kind"]
        if kind == "add":
            return f'<ins class="crit-add">{escapeHtml(meta["content"])}</ins>'
        if kind == "del":
            return f'<del class="crit-del">{escapeHtml(meta["content"])}</del>'
        if kind == "sub":
            return (
                f'<del class="crit-del">{escapeHtml(meta["old"])}</del>'
                f'<ins class="crit-add">{escapeHtml(meta["neu"])}</ins>'
            )
        if kind == "mark":
            return f'<mark class="crit-mark">{escapeHtml(meta["content"])}</mark>'
        if kind == "comment":
            return f'<span class="crit-comment">{escapeHtml(meta["content"])}</span>'
        return ""

    md.add_render_rule("cm_critic", render_critic)
